Introduce EosReformatter and an utility to run it

kvahlman opened the pull request:
EosReformatter is a reboot of the verify and write process, wrapped in
a nice package with a bow on top. Should be more efficient too, eventually.

The runner utility will allow easy testing and other image writing
activities in a more automatable fashion than the full installer.

https://phabricator.endlessm.com/T14341

kvahlman committed:
Introduce EosReformatter and an utility to run it

EosReformatter is a reboot of the verify and write process, wrapped in
a nice package with a bow on top. Should be more efficient too, eventually.

The runner utility will allow easy testing and other image writing
activities in a more automatable fashion than the full installer.

wjt commented on gnome-image-installer/util/run-reformatter.c:
```diff
@@ -0,0 +1,140 @@
+
+static gboolean _quit (gpointer data)
+{
+  g_warning (__FUNCTION__);
```
Why `g_warning`s?

kvahlman replied:
Just a debugging habit, I'll remove all of those that are not actual warnings.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,972 @@
+      gsize page_size = reformatter->page_size;
+
+      buf = g_new0(EosAlignedBuffer, 1);
+      buf->ptr = (guchar*) (((gintptr) (unaligned + page_size)) & (~(page_size - 1)));
```
I really feel like the buffer allocation etc. would be simpler if we just had a pool of `total_buffers` pointers each allocated with:

```c
if (0 != posix_memalign(&ptr, page_size, buffer_size)) { g_error("oh no"); }
```

rather than allocating a pool and then reimplementing aligned allocations from that pool.

kvahlman replied:
We still need to carry the size information though, since the decompression step produces variable-sized output (unless we make that step complex by always collecting fixed size buffers until passing it on). We could allocate the memory for the buffers with posix_memalign, though I'm not sure if that's better or worse than a large contiguous blob of memory.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,972 @@
+      /* If we can't create a new buffer, we need to wait for one */
+      EosAlignedBuffer *buf;
+
+      while (!exhaust && g_async_queue_length (reformatter->free_queue) < half_buffers);
```
I'm a bit suspicious of this. Would it not be better to factor out the buffer pool and then have two instances, one for the reader and one for the decompressor, rather than having this `exhaust` flag?

kvahlman replied:
Yes, I agree that two queues would be cleaner. This was band-aid to get the scenario of read and decompress threads being separate work properly. I think the original plan was to have just 2 threads so this wouldn't be an issue but worth fixing to use two queues I guess.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,972 @@
+      do
+        {
+          g_thread_yield();
+          buf = g_async_queue_try_pop (reformatter->free_queue);
```
Rather than spinning, could we not use a blocking pop?

kvahlman replied:
We can, if we can guarantee that it gets popped during dispose. Also the code below this appears to assume buf is non-NULL so it's a bugged logic anyway.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,972 @@
+      const gchar *type = NULL;
+
+      /* If we weren't seeded with the target size, use the file size.
+       * This is of course wrong for compressed files */
```
I don't think we should have logic which is documented to be wrong!

kvahlman replied:
The comment predates the type detection, now that we know this indeed should be turned into an error if the file is compressed. Will fix.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,972 @@
+    }
+
+  /* Initialize our pool, if it's not created yet */
+  if (reformatter->pool == NULL)
```
What's the benefit in lazily allocating the pool?

kvahlman replied:
Since the pool might be large, it would be awkward if the reformatter is created ahead of time and it would immediately gulp a large chunk of memory. Not sure how relevant the use-case is though.

wjt replied:
Doing it here seems very suspicious to me, particularly since many threads call `eos_reformatter_get_free_buffer` without any locking.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,972 @@
+
+  if (info == NULL)
+    {
+      reformatter->error = g_error_new (EOS_REFORMATTER_ERROR, 0, _("Internal error"));
```
Wouldn't it be better to propagate the error from `g_file_query_info`, possibly prefixed with some extra information about the context, and have the error page examine the error domain to work out what to tell the user (with "Internal error" being the fallback for unexpected errors)?

wjt replied:
I'm also not that wild about `reformatter->error` being set from potentially many threads without any locking.

kvahlman replied:
Yes, propagating meaningful errors is definitely the intent, the error handling was a bit rushed in to move forward. Hence the fuzzy ownership as well. We'll probably want the error to be delivered either with the finished signal or separate error signal.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,972 @@
+  if (reformatter->error != NULL)
+    {
+      /* Error is set so send EOS and abort */
+      if (buf != NULL)
```
Surely `buf == NULL` only if an error has occurred, so you can just write a `NULL` (say) to the `outq` and return.

wjt replied:
(I think `NULL` as the sentinel by which errors are signalled to consumers makes more sense than a freshly-allocated, zero-length buffer.)

kvahlman replied:
I was initially trying to avoid adding NULL-check clutter for the buffers but indeed it is a cleaner solution than allocating new buffers. will fix.

kvahlman replied:
Or not, since I now recalled that not only did I want to avoid NULL checks but also g_async_queue_push() doesn't allow NULL to be pushed.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,972 @@
+eos_reformatter_prepare_read (EosReformatter *reformatter)
+{
+  GFileInfo *info;
+  GFile *file = g_file_new_for_path (reformatter->image);
```
If you declare `info` and `file` as follows:

```c
g_autoptr(GFileInfo) info = NULL;
g_autoptr(GFile) file = NULL;
```

then they will be automatically unreffed when they fall out of scope.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,972 @@
+GError *
+eos_reformatter_get_error (EosReformatter *reformatter)
+{
+  return reformatter->error;
```
Who owns the returned `GError *`? It should either be `const`, or the API should be rethought.

kvahlman replied:
Yeah, the API should be changed to deliver the error through something better than a simple getter.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,972 @@
+      res = G_CONVERTER_CONVERTED;
+    }
+
+  if (error != NULL)
```
In the `PARTIAL_INPUT` case this branch is wrong.

kvahlman replied:
Good catch! Another bug from the error handling additions...

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,972 @@
+  EosAlignedBuffer *buf;
+  gsize wb = 0;
+
+  if (reformatter->error != NULL)
```
Maybe use a `GCancellable` to indicate the case when all the threads can stop. As a bonus this might let us add a Cancel button (of dubious utility of course…).

kvahlman replied:
Well we do have a cancel functionality already, GCancellable makes sense to add as the interface for this.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,972 @@
+  g_warning ("%s: using %d threads", __FUNCTION__, EOS_THREADS);
+
+#if EOS_THREADS == 1
+  reformatter->write_thread = g_thread_new ("read-write", do_reformat, reformatter);
```
Each thread needs to own a ref to the reformatter, I think.

kvahlman replied:
Right now we (try to) make sure the threads will exit before dispose finishes, so no explicit ref is needed. If we add refs in the threads, a simple g_object_unref() would not trigger shutdown I believe.

kvahlman committed:
Clean up debug prints, add more placeholder error reporting

kvahlman committed:
Move freeing allocations to finalize

kvahlman committed:
Rework flow control to use GCancellable, improve error handling

Also make us use 2 threads always, that seems the best option.

kvahlman committed:
Allow overriding the target device with an open file descriptor

kvahlman committed:
Add periodic syncing to keep the system from caching too much of our writes

Also do estimator sample updates with the same period, it seems to help
stabilize the update numbers a bit.

kvahlman committed:
Prevent bootable result if we abort due to errors

We zero out start of the target so the boot sector is empty if we abort
writing, and write it out as the final buffer.

kvahlman committed:
Adjust the finished signal to indicate status, and clarify error ownership.

kvahlman committed:
Enrich the error reporting with as much details as we have to give

kvahlman committed:
Remove unneeded NULL-checks around g_free()

kvahlman committed:
Don't close the write_fd if it's externally set.

Implies that after finished signal is sent, we don't touch the write_fd anymore.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,1088 @@
+  GduEstimator *estimator;
+  GConverter *decompressor;
+
+  GMutex thread_mutex;
```
I think we need to be clearer which variables are protected by this mutex.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,1088 @@
+
+  if (len < 0 || buf->len < 0)
+    {
+      g_mutex_lock (&reformatter->thread_mutex);
```
This "lock; assign to `error`; unlock" logic is duplicated many times, it should be in a function. Also: every case fails to check whether `error` is non-`NULL`.

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,1088 @@
+{
+  EosReformatter *reformatter = EOS_REFORMATTER(data);
+
+  gdu_estimator_add_sample (reformatter->estimator, reformatter->write_completed_bytes);
```
`write_completed_bytes` is modified by a different thread so this read should be guarded by the mutex too, no?

wjt commented on gnome-image-installer/util/eos-reformatter.c:
```diff
@@ -0,0 +1,1088 @@
+  g_return_val_if_fail (reformatter->image != NULL, FALSE);
+  g_return_val_if_fail (reformatter->device != NULL, FALSE);
+
+  g_object_set (reformatter, "progress", 0.0, NULL);
```
We should have some guard against this being called more than once.

timofonic commented:
@endlessm @wjt  @kvahlman  What happened to this?

wjt commented:
Superseded by #56.
